django/db/models/expressions.py

```diff
@@ -422,16 +423,161 @@ def __hash__(self):
         return hash(self.identity)
 
 
-_connector_combinators = {
-    connector: [
-        (fields.IntegerField, fields.IntegerField, fields.IntegerField),
-        (fields.IntegerField, fields.DecimalField, fields.DecimalField),
-        (fields.DecimalField, fields.IntegerField, fields.DecimalField),
-        (fields.IntegerField, fields.FloatField, fields.FloatField),
-        (fields.FloatField, fields.IntegerField, fields.FloatField),
-    ]
-    for connector in (Combinable.ADD, Combinable.SUB, Combinable.MUL, Combinable.DIV)
-}
+# Type infererence for CombinedExpression.output_field.
+# Missing items will result in FieldError, by design.
+
+# Notes:
+#
+# Different databases may handle type combinations in different ways.
+# To resolve these, we can use:
+# - What Django did in the past, for compatibility
+# - What Python normally does
+# - What makes most sense
+
+# We should also note that currently `output_field` has limited scope
+# in what it affects:
+#
+# - it does not cause a database level CAST to be done (see ticket #26650)
+# - it controls conversion to Python data types, but that doesn't necessarily
+#   include validation, so some invalid values for the field may be returned.
+
+# Correct behaviour for `NoneType` is difficult to decide, due to different
+# database handling of NULL, but we base these on lowest common denominator
+# behaviour e.g. if one supported database is going to raise an error (rather
+# than return NULL) for `val <op> NULL`, then we should raise FieldError for
+# `annotate(field=value <op> None)`. We allow overriding using
+# `ExpressionWrapper(output_field=...)` if the user knows better.
+
+NoneType = type(None)
+
+_connector_combinations = [
+    # Numeric operations - operands of same type
+    {
+        connector: [
+            (fields.IntegerField, fields.IntegerField, fields.IntegerField),
+            (fields.FloatField, fields.FloatField, fields.FloatField),
+            (fields.DecimalField, fields.DecimalField, fields.DecimalField),
+        ]
+        for connector in (
+            Combinable.ADD,
+            Combinable.SUB,
+            Combinable.MUL,
+            # Behaviour for DIV with integer arguments follows Postgres/SQLite,
+            # not MySQL/Oracle
+            Combinable.DIV,
+            Combinable.MOD,
+            Combinable.POW,
+        )
+    },
+
+    # Numeric operations - operands of different type
+    {
+        connector: [
+            (fields.IntegerField, fields.DecimalField, fields.DecimalField),
+            (fields.DecimalField, fields.IntegerField, fields.DecimalField),
+            (fields.IntegerField, fields.FloatField, fields.FloatField),
+            (fields.FloatField, fields.IntegerField, fields.FloatField),
+        ]
+        for connector in (
+            Combinable.ADD,
+            Combinable.SUB,
+            Combinable.MUL,
+            Combinable.DIV,
```

Comment:
@spookylukey I noticed that you don't have `MOD` operation here. Since you can do something like 4.5%2, is there a reason why thats not included?

Reply:
I didn't include it, because in this patch I'm trying to simply correct behaviour for date/times etc, while preserving as much of existing behaviour (that was correct) as possible.

Previous behaviour allowed MOD for operands of the same input and output type (implicitly due to logic in `_resolve_output_type`). It also added some special cases for different type for other numeric operators - ADD/SUB/MUL/DIV - but not MOD.

We could certainly add that behaviour, but I think it would be best to do it in another patch, and after checking what different databases do, in case of incompatibilities.